src/zulip/mod.rs

```diff
@@ -0,0 +1,191 @@
+                .collect::<Vec<_>>();
+            (name, member_ids)
+        })
+        .collect();
+    Ok(user_group_definitions)
+}
+
+/// Interacts with the Zulip API
+struct ZulipController {
+    /// User group name to Zulip user group id
+    user_group_ids: BTreeMap<String, ZulipUserGroup>,
+    /// The Zulip API
+    zulip_api: ZulipApi,
+    /// Whether this is a dry run or not
+    dry_run: bool,
+}
+
+impl ZulipController {
+    /// Create a new `ZulipController`
+    fn new(zulip_api: ZulipApi, dry_run: bool) -> Result<Self, Error> {
+        let user_groups = zulip_api.get_user_groups()?;
+
+        let user_group_ids = user_groups
+            .into_iter()
+            .map(|mut ug| {
+                // sort for better diagnostics
+                ug.members.sort_unstable();
+                (ug.name.clone(), ug)
+            })
+            .collect();
+
+        Ok(Self {
+            user_group_ids,
+            zulip_api,
+            dry_run,
+        })
+    }
+
+    /// Create or update a user group for the given team name and members
+    fn create_or_update_user_group(
+        &mut self,
+        user_group_name: &str,
+        member_zulip_ids: &[usize],
+    ) -> Result<(), Error> {
+        let id = self.user_group_id_from_name(&user_group_name);
+        let user_group_id = match id {
+            Some(id) => {
+                log::info!(
+                    "'{}' user group ({}) already exists on Zulip",
+                    user_group_name,
+                    id
+                );
+                id
+            }
+            None => {
+                log::info!(
+                    "no '{}' user group found on Zulip. Creating one...",
+                    user_group_name
+                );
+                self.create_user_group(
+                    &user_group_name,
+                    &format!("The {} team", user_group_name),
+                    member_zulip_ids,
+                )?;
+                return Ok(());
+            }
+        };
+
+        let existing_members = self.user_group_members_from_name(&user_group_name).unwrap();
+        log::info!(
+            "'{}' user group ({}) has members on Zulip {:?} and needs to have {:?}",
+            user_group_name,
+            user_group_id,
+            existing_members,
+            member_zulip_ids
+        );
+        let add_ids = member_zulip_ids
+            .iter()
+            .filter(|i| !existing_members.contains(i))
+            .copied()
+            .collect::<Vec<_>>();
+        let remove_ids = existing_members
+            .iter()
+            .filter(|i| !member_zulip_ids.contains(i))
+            .copied()
+            .collect::<Vec<_>>();
+
+        // We don't currently update the members field of the cached user group because it's
+        // not necessary, but for correctness sake we should consider doing so
+        self.zulip_api
+            .update_user_group_members(user_group_id, &add_ids, &remove_ids)
```

Comment:
We should probably also synchronize the description.

If we go with the "managed by the team repo" thing in the description to track groups managed by sync-team, synchronizing the description would also mark existing teams that are added on the team repo at a later point in time as managed by the team repo, which would be great.

Reply:
Sure! Doing it in a followup PR sounds fine, it's not required for the initial synchronization work.